File: legacy/src/main/rs/emulate/legacy/model/ModelDecoder.java
package rs.emulate.legacy.model;

import com.google.common.base.Preconditions;
import rs.emulate.shared.util.DataBuffer;

import java.util.EnumSet;
import java.util.HashSet;
import java.util.Set;
import java.util.function.Supplier;

import static rs.emulate.legacy.model.ModelFeature.*;

/**
 * A decoder for {@link Model}s.
 */
public final class ModelDecoder {
    /**
     * The size in {@code byte}s of the model header block.
     */
    public static final int HEADER_SIZE = 18;

    /**
     * Bitmask for vertex data blocks that contain an X coordinate.
     */
    public static final int VERTEX_X_POSITION = 0x1;

    /**
     * Bitmask for vertex data blocks that contain a Y coordinate.
     */
    public static final int VERTEX_Y_POSITION = 0x2;

    /**
     * Bitmask for vertex data blocks that contain a Z coordinate.
     */
    public static final int VERTEX_Z_POSITION = 0x4;

    /**
     * The {@link Model} data to decode.
     */
    private final DataBuffer buffer;

    /**
     * An indexed array of {@link Vertex}es in the {@link Model} being decoded.
     */
    private Vertex[] vertices;

    /**
     * An indexed array of texture coordinates projected into viewspace.
     */
    private TexCoord[] texCoords;

    /**
     * The faces of the geometry in the {@link Model} being decoded.
     */
    private Face[] faces;

    /**
     * A set of model rendering features enabled for the {@link Model} being decoded.
     */
    private Set<ModelFeature> features = new HashSet<>();

    /**
     * Create a decoder for the {@link Model} data stored in {@code buffer}.
     *
     * @param buffer The uncompressed {@link DataBuffer} containing the encoded {@code Model}.
     */
    public ModelDecoder(DataBuffer buffer) {
        this.buffer = buffer;
    }

    /**
     * Decode a feature flag value using {@code flagDecoder} and add {@code feature} to the feature set if
     * it was enabled.
     *
     * @param featureSet The set that the feature will be added to if enabled.
     * @param feature The feature to check.
     * @param flagDecoder A supplier that decodes and returns the flag.
     */
    private static void checkFeatureFlag(Set<ModelFeature> featureSet, ModelFeature feature,
                                         Supplier<Boolean> flagDecoder) {

        boolean enabled = flagDecoder.get();
        if (enabled) {
            featureSet.add(feature);
        }
    }

    /**
     * Decode the {@link ModelFeature}s present given the attributes in the {@code header} block.
     *
     * @param header The data buffer containing the model header block.
     * @return A {@link Set} of {@link ModelFeature}s that are enabled.
     */
    private Set<ModelFeature> decodeFeatures(DataBuffer header) {
        Set<ModelFeature> features = EnumSet.noneOf(ModelFeature.class);

        checkFeatureFlag(features, FACE_TEXTURES, header::getBoolean);
        checkFeatureFlag(features, FACE_RENDER_PRIORITY, () -> header.getUnsignedByte() == 255);
        checkFeatureFlag(features, FACE_TRANSPARENCY, header::getBoolean);
        checkFeatureFlag(features, FACE_SKINNING, header::getBoolean);
        checkFeatureFlag(features, VERTEX_SKINNING, header::getBoolean);

        return features;
    }

    public Model decode() {
        DataBuffer header = buffer.asReadOnlyBuffer();
        header.position(buffer.capacity() - HEADER_SIZE);

        vertices = new Vertex[header.getUnsignedShort()];
        faces = new Face[header.getUnsignedShort()];
        texCoords = new TexCoord[header.getUnsignedByte()];
        features = decodeFeatures(header);

        int xDataLength = header.getUnsignedShort();
        int yDataLength = header.getUnsignedShort();
        int zDataLength = header.getUnsignedShort();
        int faceDataLength = header.getUnsignedShort();

        int offset = 0;
        offset += vertices.length;

        int faceTypesOffset = offset;
        offset += faces.length;

        int faceRenderPrioritiesOffset = offset;
        if (features.contains(FACE_RENDER_PRIORITY)) {
            offset += faces.length;
        }

        int faceBonesOffset = offset;
        if (features.contains(FACE_SKINNING)) {
            offset += faces.length;
        }

        int faceTexturePointersOffset = offset;
        if (features.contains(FACE_TEXTURES)) {
            offset += faces.length;
        }

        int vertexBonesOffset = offset;
        if (features.contains(VERTEX_SKINNING)) {
            offset += vertices.length;
        }

        int faceAlphasOffset = offset;
        if (features.contains(FACE_TRANSPARENCY)) {
            offset += faces.length;
        }

        int faceDataOffset = offset;
        offset += faceDataLength;

        int faceColoursOffset = offset;
        offset += faces.length * 2;

        int texCoordsOffset = offset;
        offset += texCoords.length * 6;

        int xDataOffset = offset;
        offset += xDataLength;

        int yDataOffset = offset;
        offset += yDataLength;

        int zDataOffset = offset;
        offset += zDataLength;

        decodeVertices(xDataOffset, yDataOffset, zDataOffset, vertexBonesOffset);
        decodeFaces(faceDataOffset, faceTypesOffset, faceColoursOffset, faceRenderPrioritiesOffset,
            faceAlphasOffset, faceBonesOffset, faceTexturePointersOffset);
        decodeTexCoords(texCoordsOffset);

        return new Model(features, faces, vertices, texCoords);
    }

    /**
     * Decode the projected texture coordinates for the {@link Model} being decoded.
     *
     * @param offset The offset into {@code buffer} of the texture coordinate data.
     */
    private void decodeTexCoords(int offset) {
        DataBuffer texCoordsBuffer = buffer.asReadOnlyBuffer();
        texCoordsBuffer.position(offset);

        for (int index = 0; index < texCoords.length; index++) {
            int origin = texCoordsBuffer.getUnsignedByte();
            int u = texCoordsBuffer.getUnsignedByte();
            int v = texCoordsBuffer.getUnsignedByte();
            texCoords[index] = new TexCoord(origin, u, v);
        }
    }

    /**
     * Decode the triangle {@link Face}s that make up the geometry of the {@link Model} being decoded.
     *
     * @param faceDataOffset The index into {@code buffer} of the face geometry data.
     * @param typesOffset The index into {@code buffer} of the face type data.
     * @param coloursOffset The index into {@code buffer} of the face colour data.
     * @param renderPrioritiesOffset The index into {@code buffer} of the render priority data.
     * @param alphasOffset The index into {@code buffer} of the alpha value data.
     * @param bonesOffset The index into {@code buffer} of the face skinning data.
     * @param texturePointersOffset The index into {@code buffer} of the the texture pointer data.
     */
    private void decodeFaces(int faceDataOffset, int typesOffset,
                             int coloursOffset, int renderPrioritiesOffset,
                             int alphasOffset, int bonesOffset, int texturePointersOffset) {
        Preconditions.checkState(faces.length > 0, "Must be 1 or more faces present");

        DataBuffer faceData = buffer.asReadOnlyBuffer();
        faceData.position(faceDataOffset);

        DataBuffer types = buffer.asReadOnlyBuffer();
        types.position(typesOffset);

        DataBuffer colours = buffer.asReadOnlyBuffer();
        colours.position(coloursOffset);

        DataBuffer renderPriorities = buffer.asReadOnlyBuffer();
        renderPriorities.position(renderPrioritiesOffset);

        DataBuffer alphas = buffer.asReadOnlyBuffer();
        alphas.position(alphasOffset);

        DataBuffer bones = buffer.asReadOnlyBuffer();
        bones.position(bonesOffset);

        DataBuffer texturePointers = buffer.asReadOnlyBuffer();
        texturePointers.position(texturePointersOffset);

        int faceA = 0;
        int faceB = 0;
        int faceC = 0;
        int offset = 0;

        for (int index = 0; index < faces.length; index++) {
            int type = types.getUnsignedByte();
            int colour = colours.getUnsignedShort();
            int renderPriority = features.contains(FACE_RENDER_PRIORITY) ? renderPriorities.getUnsignedByte() : -1;
            int alpha = features.contains(FACE_TRANSPARENCY) ? alphas.getUnsignedByte() : -1;
            int bone = features.contains(FACE_SKINNING) ? bones.getUnsignedByte() : -1;
            int texturePointer = features.contains(FACE_TEXTURES) ? bones.getUnsignedByte() : -1;

            if (type == 1) {
                faceA = faceData.getSignedSmart() + offset;
                offset = faceA;
                faceB = faceData.getSignedSmart() + offset;
                offset = faceB;
                faceC = faceData.getSignedSmart() + offset;
                offset = faceC;
            } else if (type == 2) {
                faceB = faceC;
                faceC = faceData.getSignedSmart() + offset;
                offset = faceC;
            } else if (type == 3) {
                faceA = faceC;
                faceC = faceData.getSignedSmart() + offset;
                offset = faceC;
            } else if (type == 4) {
                int temp = faceA;
                faceA = faceB;
                faceB = temp;

                faceC = faceData.getSignedSmart() + offset;
                offset = faceC;
            }

            faces[index] = new Face(faceA, faceB, faceC, colour, renderPriority, alpha, bone, texturePointer);
        }
    }

    /**
     * Decode a list of {@link Vertex} objects from the input {@link Model} file.
     *
     * @param xDataOffset the offset into {@link this#buffer} of the X coordinate buffer.
     * @param yDataOffset the offset into {@link this#buffer} of the Y coordinate buffer.
     * @param zDataOffset the offset into {@link this#buffer} of the Z coordinate bfufer.
     */
    private void decodeVertices(int xDataOffset, int yDataOffset, int zDataOffset, int vertexBonesOffset) {
        Preconditions.checkState(vertices.length > 0, "Vertex count must be greater than 0");

        DataBuffer directionBuffer = buffer.asReadOnlyBuffer();
        directionBuffer.position(0);
        DataBuffer verticesX = buffer.asReadOnlyBuffer();
        verticesX.position(xDataOffset);

        DataBuffer verticesY = buffer.asReadOnlyBuffer();
        verticesY.position(yDataOffset);

        DataBuffer verticesZ = buffer.asReadOnlyBuffer();
        verticesZ.position(zDataOffset);

        DataBuffer bones = buffer.asReadOnlyBuffer();
        bones.position(vertexBonesOffset);

        int baseX = 0;
        int baseY = 0;
        int baseZ = 0;

        for (int index = 0; index < vertices.length; index++) {
            int mask = directionBuffer.getUnsignedByte();
            int x = 0;
            if ((mask & VERTEX_X_POSITION) != 0) {
                x = verticesX.getSignedSmart();
            }

            int y = 0;
            if ((mask & VERTEX_Y_POSITION) != 0) {
                y = verticesY.getSignedSmart();
            }

            int z = 0;
            if ((mask & VERTEX_Z_POSITION) != 0) {
                z = verticesZ.getSignedSmart();
            }

            x += baseX;
            y += baseY;
            z += baseZ;
            baseX = x;
            baseY = y;
            baseZ = z;

            vertices[index] = new Vertex(x, y, z);
            //@todo
            int bone = features.contains(VERTEX_SKINNING) ? bones.getUnsignedByte() : -1;
        }
    }
}

File: legacy/src/main/kotlin/rs/emulate/legacy/model/ModelDecoder.kt
package rs.emulate.legacy.model

import io.netty.buffer.ByteBuf
import rs.emulate.legacy.model.ModelFeature.*
import rs.emulate.util.readSignedSmart
import java.util.HashSet

/**
 * A decoder for [Model]s.
 *
 * @param buffer The uncompressed [ByteBuf] containing the encoded `Model`.
 */
class ModelDecoder(private val buffer: ByteBuf) {

    /**
     * An indexed array of [Vertex]es in the [Model] being decoded.
     */
    private lateinit var vertices: Array<Vertex?>

    /**
     * An indexed array of texture coordinates projected into viewspace.
     */
    private lateinit var texCoords: Array<TexCoord?>

    /**
     * The faces of the geometry in the [Model] being decoded.
     */
    private lateinit var faces: Array<Face?>

    /**
     * A set of model rendering features enabled for the [Model] being decoded.
     */
    private var features: Set<ModelFeature> = HashSet()

    fun decode(): Model {
        val header = buffer.copy()
        header.readerIndex(buffer.capacity() - HEADER_SIZE)

        vertices = arrayOfNulls(header.readUnsignedShort())
        faces = arrayOfNulls(header.readUnsignedShort())
        texCoords = arrayOfNulls(header.readUnsignedByte().toInt())

        features = ModelFeature.decodeFrom(header)

        val xDataLength = header.readUnsignedShort()
        val yDataLength = header.readUnsignedShort()
        val zDataLength = header.readUnsignedShort()
        val faceDataLength = header.readUnsignedShort()

        var offset = vertices.size

        val faceTypesOffset = offset
        offset += faces.size

        val faceRenderPrioritiesOffset = offset
        if (features.none { it is GlobalFaceRenderPriority }) { // TODO full scan is stupid
            offset += faces.size
        }

        val faceBonesOffset = offset
        if (FaceSkinning in features) {
            offset += faces.size
        }

        val faceTexturePointersOffset = offset
        if (FaceTextures in features) {
            offset += faces.size
        }

        val vertexBonesOffset = offset
        if (VertexSkinning in features) {
            offset += vertices.size
        }

        val faceAlphasOffset = offset
        if (FaceTransparency in features) {
            offset += faces.size
        }

        val faceDataOffset = offset
        offset += faceDataLength

        val faceColoursOffset = offset
        offset += faces.size * 2

        val texCoordsOffset = offset
        offset += texCoords.size * 6

        val xDataOffset = offset
        offset += xDataLength

        val yDataOffset = offset
        offset += yDataLength

        val zDataOffset = offset
        offset += zDataLength

        decodeVertices(xDataOffset, yDataOffset, zDataOffset, vertexBonesOffset)
        decodeFaces(faceDataOffset, faceTypesOffset, faceColoursOffset, faceRenderPrioritiesOffset,
            faceAlphasOffset, faceBonesOffset, faceTexturePointersOffset)
        decodeTexCoords(texCoordsOffset)

        return Model(features, faces.requireNoNulls(), vertices.requireNoNulls(), texCoords.requireNoNulls())
    }

    /**
     * Decode the projected texture coordinates for the [Model] being decoded.
     *
     * @param offset The offset into `buffer` of the texture coordinate data.
     */
    private fun decodeTexCoords(offset: Int) {
        val texCoordsBuffer = buffer.copy()
        texCoordsBuffer.readerIndex(offset)

        for (index in texCoords.indices) {
            val origin = texCoordsBuffer.readUnsignedByte().toInt()
            val u = texCoordsBuffer.readUnsignedByte().toInt()
            val v = texCoordsBuffer.readUnsignedByte().toInt()

            texCoords[index] = TexCoord(origin, u, v)
        }
    }

    /**
     * Decode the triangle [Face]s that make up the geometry of the [Model] being decoded.
     */
    private fun decodeFaces(
        faceDataOffset: Int,
        typesOffset: Int,
        coloursOffset: Int,
        renderPrioritiesOffset: Int,
        alphasOffset: Int,
        bonesOffset: Int,
        texturePointersOffset: Int
    ) {
        check(faces.isNotEmpty()) { "Must be 1 or more faces present." }

        val faceData = buffer.copy()
        faceData.readerIndex(faceDataOffset)

        val types = buffer.copy()
        types.readerIndex(typesOffset)

        val colours = buffer.copy()
        colours.readerIndex(coloursOffset)

        val priorities = buffer.copy()
        priorities.readerIndex(renderPrioritiesOffset)

        val alphas = buffer.copy()
        alphas.readerIndex(alphasOffset)

        val bones = buffer.copy()
        bones.readerIndex(bonesOffset)

        val texturePointers = buffer.copy()
        texturePointers.readerIndex(texturePointersOffset)

        var faceA = 0
        var faceB = 0
        var faceC = 0
        var offset = 0

        val priorityFeature = features.find { it is GlobalFaceRenderPriority } as GlobalFaceRenderPriority?

        for (index in faces.indices) {
            val type = types.readUnsignedByte().toInt()
            val colour = colours.readUnsignedShort()

            val renderPriority = priorityFeature?.priority ?: priorities.readUnsignedByte().toInt()

            val alpha = if (FaceTransparency in features) alphas.readUnsignedByte().toInt() else -1
            val bone = if (FaceSkinning in features) bones.readUnsignedByte().toInt() else -1
            val texturePointer = if (FaceTextures in features) bones.readUnsignedByte().toInt() else -1

            if (type == 1) {
                faceA = faceData.readSignedSmart() + offset
                offset = faceA

                faceB = faceData.readSignedSmart() + offset
                offset = faceB

                faceC = faceData.readSignedSmart() + offset
                offset = faceC
            } else if (type == 2) {
                faceB = faceC
                faceC = faceData.readSignedSmart() + offset
                offset = faceC
            } else if (type == 3) {
                faceA = faceC
                faceC = faceData.readSignedSmart() + offset
                offset = faceC
            } else if (type == 4) {
                val temp = faceA
                faceA = faceB
                faceB = temp

                faceC = faceData.readSignedSmart() + offset
                offset = faceC
            }

            faces[index] = Face(faceA, faceB, faceC, colour, renderPriority, alpha, bone, texturePointer)
        }
    }

    /**
     * Decode a list of [Vertex] objects from the input [Model] file.
     */
    private fun decodeVertices(xDataOffset: Int, yDataOffset: Int, zDataOffset: Int, vertexBonesOffset: Int) {
        check(vertices.isNotEmpty()) { "Vertex count must be greater than 0." }

        val directionBuffer = buffer.copy()
        directionBuffer.readerIndex(0)
        val verticesX = buffer.copy()
        verticesX.readerIndex(xDataOffset)

        val verticesY = buffer.copy()
        verticesY.readerIndex(yDataOffset)

        val verticesZ = buffer.copy()
        verticesZ.readerIndex(zDataOffset)

        val bones = buffer.copy()
        bones.readerIndex(vertexBonesOffset)

        var baseX = 0
        var baseY = 0
        var baseZ = 0

        for (index in vertices.indices) {
            val mask = directionBuffer.readUnsignedByte().toInt()
            var x = when {
                mask and VERTEX_X_POSITION != 0 -> verticesX.readSignedSmart()
                else -> 0
            }

            var y = when {
                mask and VERTEX_Y_POSITION != 0 -> verticesY.readSignedSmart()
                else -> 0
            }

            var z = when {
                mask and VERTEX_Z_POSITION != 0 -> verticesZ.readSignedSmart()
                else -> 0
            }

            val bone = when (VertexSkinning) {
                in features -> bones.readUnsignedByte().toInt()
                else -> -1
            }

            x += baseX
            y += baseY
            z += baseZ

            baseX = x
            baseY = y
            baseZ = z

            vertices[index] = Vertex(x, y, z, bone)
        }
    }

    companion object {

        /**
         * The size (in bytes) of the model header block.
         */
        private const val HEADER_SIZE = 18

        /**
         * The mask for vertex data blocks that contain an X coordinate.
         */
        private const val VERTEX_X_POSITION = 0x1

        /**
         * The mask for vertex data blocks that contain a Y coordinate.
         */
        private const val VERTEX_Y_POSITION = 0x2

        /**
         * The mask for vertex data blocks that contain a Z coordinate.
         */
        private const val VERTEX_Z_POSITION = 0x4

    }
}
